src/docfx/build/redirection/BuildRedirection.cs

```diff
@@ -18,23 +18,17 @@ internal static List<Error> Build(Context context, Document file)
             var (monikerError, monikers) = context.MonikerProvider.GetFileLevelMonikers(file.FilePath);
             errors.AddIfNotNull(monikerError);
 
-            var publishItem = new PublishItem
-            {
-                Url = file.SiteUrl,
-                SourcePath = file.FilePath.Path,
-                Locale = context.LocalizationProvider.Locale,
-                RedirectUrl = context.RedirectionProvider.GetRedirectUrl(file.FilePath),
-                Monikers = monikers,
-                MonikerGroup = MonikerUtility.GetGroup(monikers),
-                ConfigMonikerRange = context.MonikerProvider.GetConfigMonikerRange(file.FilePath),
-            };
-
-            if (context.Config.Legacy)
-            {
-                publishItem.Path = context.DocumentProvider.GetOutputPath(file.FilePath, monikers);
-            }
+            var publishItem = new PublishItem(
+                file.SiteUrl,
+                context.Config.Legacy ? context.DocumentProvider.GetOutputPath(file.FilePath, monikers) : null,
+                file.FilePath.Path,
+                context.LocalizationProvider.Locale,
+                monikers,
+                context.MonikerProvider.GetConfigMonikerRange(file.FilePath));
+
+            publishItem.RedirectUrl = context.RedirectionProvider.GetRedirectUrl(file.FilePath);
```

Comment:
why can not `RedirectUrl` be in the constructor as well?

Reply:
It is not mandatory, same as other properties like Metadata, and in cases like metadata, it needs to be set after the publish item is created.